pkg/timelock/timelock.go

```diff
@@ -128,166 +128,256 @@ func NewTimelockWorker(nodeURL, timelockAddress, callProxyAddress, privateKey st
+						sub, err = tw.ethClient.SubscribeFilterLogs(ctx, query, logCh)
+						if err == nil {
+							tw.logger.Info().Msg("subscription successfully recreated.")
+							SetReadyStatus(HealthStatusOK)
+							success = true
+							break
+						}
+
+						time.Sleep(time.Second * time.Duration(try))
+					}
+
+					if !success {
+						tw.logger.Error().Msg("failed to recreate subscription after retries: shutting down timelock-worker.")
+						wg.Done()
+						return
+					}
+				}
+
+			case <-ctx.Done():
+				tw.logger.Debug().Msgf("shutting down subscription")
+				SetReadyStatus(HealthStatusError)
+				return
+			}
+		}
+	}()
 
-	// Read events by FilterLogs. This method calls eth_getLogs under the hood.
+	return nil
+}
+
+func (tw *Worker) retrieveHistoricalLogs(ctx context.Context) (chan types.Log, error) {
+	query := tw.setupFilterQuery()
+	logCh := make(chan types.Log)
+
+	// FilterLogs starts filtering the logs at fromBlock, gathering the historical data.
+	// This is needed to guarantee that all the events are gathered, even in scenarios where the worker crashes.
 	filter, err := tw.ethClient.FilterLogs(ctx, query)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	// Process incoming historical logs in a separate goroutine.
+	wg.Add(1)
 	go func() {
-		for _, l := range filter {
-			logCh <- l
+		defer wg.Done()
+		for _, log := range filter {
+			select {
+			case logCh <- log:
+				tw.logger.Debug().Msgf("processing historical log: %v\n", log)
+			case <-ctx.Done():
+				tw.logger.Debug().Msg("shutting down the retrieval of old logs in incomplete status")
+				return
+			}
 		}
+		tw.logger.Debug().Msg("retrieved correctly all historical events")
+		close(logCh)
 	}()
 
-	// Setting readyStatus here because we want to make sure subscription is up.
-	tw.logger.Info().Msgf("Initial subscription complete")
-	SetReadyStatus(HealthStatusOK)
+	return logCh, nil
+}
 
-	// This is the goroutine watching over the subscription.
-	// We want wg.Done() to cancel the whole execution, so don't add more than 1 to wg.
-	// Also, when receiving an event that creates an error, skip the event and
-	// continue processing the rest, as an external operator can cancel the faulty event.
-	loop := true
+func (tw *Worker) processLogs(ctx context.Context, oldLog, newLog chan types.Log, stopCh chan string, cancel context.CancelFunc) {
+	// This is the goroutine watching over the subscribed and historical logs.
 	wg.Add(1)
 	go func() {
-		for loop {
+		defer wg.Done()
+		for {
 			select {
-			case log := <-logCh:
-				// Decode the log into an event using the ABI exposed in Timelock.go
-				event, err := tw.ABI.EventByID(log.Topics[0])
-				if err != nil {
-					continue
+			case log := <-newLog:
+				if err := tw.handleLog(ctx, log); err != nil {
+					tw.logger.Error().Msgf("error processing new log: %v\n", log)
 				}
 
-				if event == nil {
-					continue
+			case log := <-oldLog:
```

Comment:
does order not matter?  seems you would want to process all the historical logs first and then process new logs

Reply:
no, order in select does not matter. We want to process two channels and the order is not really important.
The scheduler will handle the operations and execute them; andd the operations with a parent will fail until the parent is executed.